ocaml/xapi/xapi_xenops.ml

```diff
@@ -1956,18 +1955,19 @@ let update_vusb ~__context (id: (string * string)) =
         else begin
           let pusb, pusb_r =
             Db.VM.get_VUSBs ~__context ~self:vm
-            |> List.map (fun vusb -> Db.VUSB.get_attached ~__context ~self:vusb)
-            |> List.filter (fun attached -> Db.is_valid_ref __context attached)
+            |> List.map (fun self -> Db.VUSB.get_USB_group ~__context ~self)
+            |> List.map (fun usb_group -> List.hd (Db.USB_group.get_PUSBs ~__context ~self:usb_group))
             |> List.map (fun self -> self, Db.PUSB.get_record ~__context ~self)
             |> List.find (fun (_, pusbr) -> "vusb" ^ pusbr.API.pUSB_path= (snd id))
           in
-          let vusb = Db.PUSB.get_attached ~__context ~self:pusb in
+          let usb_group = Db.PUSB.get_USB_group ~__context ~self:pusb in
+          let vusb = List.hd (Db.USB_group.get_VUSBs ~__context ~self:usb_group) in
```

Comment:
This convoluted logic is here because we haven't got a better way of identifying which VUSB the update is associated with. Ideally we'd have something like the `device_number` that we have for VIFs - ie, a thing associated with the VUSB rather than the PUSB.

Since this is in the events_from_xenopsd code, let's catch and log exceptions here, but carry on.

Reply:
update, will raise internal error here ok?